doc/design/upgrade.md

```diff
@@ -125,15 +135,114 @@ Here is the upgrade workflow
+
+This feature can be delayed until there is new secret key added.
+
+### Container Guestinfo
+
+VIC should not force user to upgrade containers after upgrade appliance, which means appliance will need to talk with old version's containers. Read back container configuration will be similiar to appliance configuration upgrade process. Corresponding data migration plugin should be prepared for each change, and registered to upgrade framework with release version. The process is listed below.
+
+- One ExecutorConfig structure per one version, where the container configuration is changed. If no changes, that version will be skipped.
+- Portlayer will check configuration version first while reading configuration. If it's older version, run in memory data migration through upgrade framework.
+- Together with upgrade configuration, portlayer need to know if configuration is migrated or not.
+- If no data migration, continue same to new version's container
+- If has data migration, do not write back configuration (old version's container is running in readonly mode)
+
+If container data migration happens, some of docker functions, e.g. start/attach/log, and docker ps/inspect information might be impacted, but docker image/volume/stop/rm should still work well, to make sure user can destroy and recreate container easily.
+
+During appliance upgrade, vic-machine should have clear information to mention the container functional limitation after upgrade, and the solution for it.
+
+### KeyValue Store Migration
+Different to guestinfo object, VIC will not change configuration from guestinfo to anywhere else anytime soon. But from start, keyvalue store persistent position is in argument. Right now, vic has a few datastore files for keyvalue store, but those information is not configurable, which means hardcoded in portlayer.
+
+In the future, if there is any change on assumption no matter for what reason, performance, easy to management, data migration will be hard to do.
+
+So the idea is that, data migration framework does not assume where to load keyvalue store. If there is keyvalue store changes, plugin is the owner to read old version, update to new version and then write back to datastore files, or to anywhere defined in new version, suppose the configuration can be retrieved from guestinfo. So the input of keyvalue store plugin will be same to other appliance configuration update plugin.
+
+The problem of this solution is that each single plugin will persist its own change, not like the guestinfo update, which is migrated in memory and persisted by migration framework. So the rollback for guestinfo is easy, but not possible for keyvalue store plugin, unless we have data roll back plugin mechanism.
+
+The workaround for this idea is to restrict keyvalue store migration functions, e.g. only new key is supported. In the plugin, do not remove old keyvalues, so in the above failure case, old binary can still work with old keyvalues.
```

Comment:
I agree with you, and thanks for find it. We've been talked about that while you're in Austin.
I actually updated the proto type and developer guide based on this idea. But unfortunately did not find here has this description as well. Need more linkage for this kind of change.